plugins/module_utils/foreman_helper.py

```diff
@@ -329,7 +329,14 @@ def __init__(self, **kwargs):
         super(ForemanAnsibleModule, self).__init__(argument_spec=argument_spec, supports_check_mode=supports_check_mode, **kwargs)
 
         aliases = {alias for arg in argument_spec.values() for alias in arg.get('aliases', [])}
-        self.foreman_params = {k: v for (k, v) in self.params.items() if v is not None and k not in aliases}
+        self.foreman_params = {}
+        for (k, v) in self.params.items():
+            if v is not None and k not in aliases:
+                if isinstance(v, dict):
+                    v = {subk: subv for (subk, subv) in v.items() if subv is not None}
+                elif isinstance(v, list) and v and isinstance(v[0], dict):
+                    v = [{subk: subv for (subk, subv) in element.items() if subv is not None} for element in v]
+                self.foreman_params[k] = v
```

Comment:
Should this be fully recursive again?

Reply:
I honestly don't expect further nesting, but we can.

Reply:
I bet the apipie specs allow it. And maybe it will spare you an hour of debugging once `foreman-weird-plugin` uses it.